Archive/Hairer.lean

```diff
@@ -0,0 +1,242 @@
+  · refine span_le.mpr ?_ h
+    rintro x ⟨c, hc, rfl⟩
+    rw [SetLike.mem_coe, mem_restrictTotalDegree]
+    cases subsingleton_or_nontrivial R
+    · rw [Subsingleton.elim ((fun c ↦ monomial c 1) c) 0, totalDegree_zero]; apply zero_le
+    · rw [totalDegree_monomial _ one_ne_zero]; exact hc
+
+/- Can be obtained by combining `LinearMap.proj` and `MvPolynomial.evalₗ`. -/
+/-- The evaluation of a multivariate polynomial at a point, as a linear map. -/
+def evalAtₗ {R σ : Type*} [CommSemiring R] (x : σ → R) : MvPolynomial σ R →ₗ[R] R where
+  toFun P := eval x P
+  map_add' := by simp
+  map_smul' := by simp
+
+/- what we actually want -/
+lemma analyticOn_eval {R σ : Type*} [Fintype σ] [IsROrC R] (P : MvPolynomial σ R) :
+    AnalyticOn R (fun x ↦ eval (EuclideanSpace.equiv σ R x) P) univ := fun x _ ↦ by
+  apply P.induction_on (fun r ↦ ?_) (fun p q hp hq ↦ ?_) fun p i hp ↦ ?_ -- refine doesn't work
+  · simp_rw [eval_C]; exact analyticAt_const
+  · simp_rw [eval_add]; exact hp.add hq
+  · simp_rw [eval_mul, eval_X]; exact hp.mul ((ContinuousLinearMap.proj (R := R) i).analyticAt x)
+
+lemma finite_stuff' [Finite σ] (N : ℕ) : {s : Multiset σ | Multiset.card s ≤ N}.Finite := by
+  classical
+  have := Fintype.ofFinite σ
+  let S := N • (Finset.univ.val : Multiset σ)
+  apply Finset.finite_toSet (Multiset.toFinset (Multiset.powerset S)) |>.subset
+  intro s hs
+  rw [Set.mem_setOf] at hs
+  rw [Finset.mem_coe, Multiset.mem_toFinset, Multiset.mem_powerset, Multiset.le_iff_count]
+  intro x
+  simp only [S, Multiset.count_nsmul, Multiset.count_univ, mul_one]
+  exact le_trans (s.count_le_card x) hs
+
+lemma finite_stuff [Finite σ] (N : ℕ) : {s : σ →₀ ℕ | s.sum (fun _ e ↦ e) ≤ N}.Finite := by
+  classical
+  change {s : σ →₀ ℕ | s.sum (fun _ => id) ≤ N}.Finite
+  simp only [← Finsupp.card_toMultiset]
+  refine Set.Finite.of_finite_image ?_ (Multiset.toFinsupp.symm.injective.injOn _)
+  convert finite_stuff' (σ := σ) N
+  ext x
+  rw [← AddEquiv.coe_toEquiv, Set.mem_image_equiv]
+  simp
+
+instance [Finite σ] : Module.Finite R (restrictTotalDegree σ R n) := by
+  rw [Module.finite_def, fg_top, restrictTotalDegree_eq_span]
+  exact Submodule.fg_span ((finite_stuff _).image _)
+
+end missing_polynomial
+
+section missing_linear_algebra
+
+open Module Submodule BigOperators
+
+variable {K V V' ι : Type*} [Field K] [AddCommGroup V] [Module K V] [AddCommGroup V'] [Module K V']
+   {B : V' →ₗ[K] Dual K V} {m : ι → V'}
+
+lemma flip_surj_of_inj (hB : Injective B) [FiniteDimensional K V'] : Surjective B.flip := by
+  rw [← LinearMap.range_eq_top]
+  apply Submodule.eq_top_of_finrank_eq
+  set W : Subspace K _ := LinearMap.range B.flip
+  have := W.finrank_add_finrank_dualCoannihilator_eq
+  rw [Subspace.dual_finrank_eq, ← this, eq_comm, add_right_eq_self, finrank_eq_zero, eq_bot_iff]
+  intro x hx
+  apply hB
+  ext v
+  rw [Submodule.mem_dualCoannihilator] at hx
+  simpa using hx _ (LinearMap.mem_range_self _ v)
```

Comment:
This is now in as [flip_injective_iff₁](https://github.com/leanprover-community/mathlib4/pull/8820/files#diff-1a470a10fd05f0eedb570bfbbe37c36b1a2472e9dc4c78aa7fc070d33ec2c2d2R1540-R1541)